### PR TITLE
calib: add accelerometer orthogonality calibration

### PR DESCRIPTION
**Pull request by niewim19**
## Description

This change adds `accorth` calibration type which should remove nonorthogonalities scaling vectors and static manufacture biases of accelerometer. This calibration is also morphed with `accrot` and the current `accrot` is removed.

`accorth` utilizes ellipsoid fitting via `lma` module and it does it twice: once for rough calibration of accelerometer ellipsoid parameters and later with precise measurements it corrects these parameters with a secondary accelerometer ellipsoid calibration.

In the end the accelerometer rotation is done by measuring two positions very precisely.

`accorth` was added to `calib`/`sensc` and `calibtool` modules.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: zturn drone.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: calib/accorth.c [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+	if (lma_init(3, 12, n, ellcal_lmaJacobian, ellcal_lmaResiduum, ellcal_lmaGuess, &lma) < 0) {
+		fprintf(stderr, "%s: failed to init LMA", ACCORTH_TAG);
+		return -1;
+	}
+
+	/* Write measurements to lma `samples` matrix */
+	for (i = 0; i < n; i++) {
+		MATRIX_DATA(&lma.samples, i, 0) = accorth_common.meas[i].x;
+		MATRIX_DATA(&lma.samples, i, 1) = accorth_common.meas[i].y;
+		MATRIX_DATA(&lma.samples, i, 2) = accorth_common.meas[i].z;
```
Comment (niewim19) [inline calib/accorth.c:321]: It should be:
```suggestion
		MATRIX_DATA(&lma.samples, i, 0) = data[i].x;
		MATRIX_DATA(&lma.samples, i, 1) = data[i].y;
		MATRIX_DATA(&lma.samples, i, 2) = data[i].z;
```
corrected in an additional commit on this branch

File: calib/ellcal.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+
+/*
+* Jacobian is calculated numerically by calculating differences of residuum after step in parameters.
+* Steps are taken for each parameter at a time.
+*/
+int ellcal_lmaJacobian(const matrix_t *P, const matrix_t *V, matrix_t *J, bool log)
+{
+	float dataPtmp[12] = { 0 };
```
Comment (kemonats) [inline calib/ellcal.c:82]: Unnecessary zeroing.
```suggestion
	float dataPtmp[12];
```

File: libs/calib/calib.c [hunk 3]
```diff
@@ -250,168 +250,241 @@ static int calib_magironRead(FILE *file, calib_data_t *cal)
...
-	cal->params.motlin.motorEq[1][0] = 1.031400;
-	cal->params.motlin.motorEq[1][1] = 0.085204;
+	if (strlen(paramName) != 3) {
+		return -1;
+	}
 
-	cal->params.motlin.motorEq[2][0] = 1.003427;
-	cal->params.motlin.motorEq[2][1] = 0.142546;
+	row = (uint8_t)(paramName[1] - '0'); /* convert character to unsigned int digit */
+	col = (uint8_t)(paramName[2] - '0'); /* convert character to unsigned int digit */
```
Comment (kemonats) [inline libs/calib/calib.c:265]: `row` and `col` are unsigned int type.
```suggestion
	row = paramName[1] - '0'; /* convert character to unsigned int digit */
	col = paramName[2] - '0'; /* convert character to unsigned int digit */
```

File: libs/calib/calib.c [hunk 4]
```diff
@@ -250,168 +250,241 @@ static int calib_magironRead(FILE *file, calib_data_t *cal)
...
 
-	cal->params.motlin.motorEq[3][0] = 0.996573;
-	cal->params.motlin.motorEq[3][1] = 0.137454;
-}
+	/* matrix type get through character check */
+	switch (paramName[0]) {
+		case ACC_CHAR_ORTHO:
+			if (row > (unsigned int)'9' || col > (unsigned int)'9') {
```
Comment (kemonats) [inline libs/calib/calib.c:270]: You compare to 0x39. It is ok?
Same below.
```suggestion
			if (row > 9 || col > 9) {
```

File: libs/calib/calib.c [hunk 5]
```diff
@@ -250,168 +250,241 @@ static int calib_magironRead(FILE *file, calib_data_t *cal)
...
+{
+	int motor, param;
+
+	if (strlen(paramName) != 4) {
 		return -1;
 	}
 
-	qParam = (uint8_t)(paramName[4] - '0'); /* convert character to unsigned int digit */
+	motor = (int8_t)(paramName[2] - '0'); /* convert character to unsigned int digit */
+	param = (int8_t)(paramName[3] - 'a'); /* convert character to unsigned int digit, expected range [a, b] -> [0, 1] */
```
Comment (kemonats) [inline libs/calib/calib.c:440]: `motor` and `param` are int, in comment is unsigned int.
`motor` and `param` can be negative?
If you declare `motor` and `param` as `unsigned int` then you don't have to check if they are negative.

File: libs/calib/calib.c (hunk 3 above)
Comment (niewim19) [reply libs/calib/calib.c:265]: This has to do with MISRA rule 10.4 that does not allow for implicit casting of signed values to unsigned values and explicit casting should be used instead.

File: libs/calib/calib.c (hunk 5 above)
Comment (niewim19) [reply libs/calib/calib.c:440]: Yeah, you are right. I will change that. Thanks!

File: libs/calib/calib.c (hunk 4 above)
Comment (niewim19) [reply libs/calib/calib.c:270]: Oh... yeah that's bad. This exact construction is used elsewhere throughout the file. I will correct all instances. Thanks for catching it!

File: calib/accorth.c [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+	* Body frame placed on flat surface should provide a measurement of acceleration pointing upward.
+	* Using NED frame of reference means that earth acceleration is parallel to NED z versor, but of opposite direction.
+	* Taking negative earth acceleration as NED z versor.
+	*/
+	printf("Place drone precisely horizontal and press [Enter]");
+	fflush(stdout);
+	fflush(stdin);
+	getchar();
+
+	sleep(1); /* Sleep for any vibration to disperse */
```
Comment (agkaminski) [inline calib/accorth.c:195]: Good candidate for a function (e.g. `accorth_prompt(const char *prompt)`)

File: calib/accorth.c [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+
+/* Calculating S3 and H3 from S1, S2, H1, H2. Returns -1 on calculation error, 0 otherwise. */
+static int accorth_paramsCombine(const matrix_t *S1, const matrix_t *H1, const matrix_t *S2, const matrix_t *H2, matrix_t *S3, matrix_t *H3)
+{
+	float invBuf[18], dataTmp[3];
+	matrix_t tmp = { .data = dataTmp, .rows = 3, .cols = 1, .transposed = 0 };
+
+	/* Using S3 as temporary storage for inv(S1) */
+	if (matrix_inv(S1, S3, invBuf, 18) < 0) {
```
Comment (agkaminski) [inline calib/accorth.c:265]:
```suggestion
	if (matrix_inv(S1, S3, invBuf, sizeof(invBuf) / sizeof(*invBuf)) < 0) {
```

File: calib/ellcal.c [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+	matrix_sub(&X, &H, NULL);   /* apply offset */
+	matrix_prod(&S, &X, &tmpX); /* apply ellipsoid deformation matrix */
+
+	/* calculate length of produced vector */
+	for (i = 0; i < 3; i++) {
+		r += MATRIX_DATA(&tmpX, i, 0) * MATRIX_DATA(&tmpX, i, 0);
+	}
+
+	/* Length difference between transformed vector and a unit sphere radius is the residuum */
+	*res = (float)(sqrt(r) - 1);
```
Comment (agkaminski) [inline calib/ellcal.c:70]: `sqrtf`?

**Review by agkaminski — APPROVED** [review]
Nitpicks only, LGTM

**Comment by niewim19** [comment]
All remarks implemented

**Review by agkaminski — APPROVED** [review]

**Pull request merged**